src/compose/app.ts

```diff
@@ -565,9 +565,13 @@ export class App {
 		) {
 			return generateStep('updateMetadata', { current, target });
 		} else if (target.config.running !== current.config.running) {
-			if (target.config.running) {
+			// Since the running states don't match that means they are inversed!
+			// We can achieve the target running state by doing the opposite then of what is current.
+			// However, for starting a container we must first check the restart policy.
+			if (!current.config.running && target.config.restart === 'always') {
+				// Only start the container if is stopped and wants to be restarted
 				return generateStep('start', { target });
-			} else {
+			} else if (current.config.running) {
 				return generateStep('stop', { current });
 			}
```

Comment:
I'd change this slightly to be
```ts
			if (current.config.running) {
				// If it's running and should be stopped then stop it
				return generateStep('stop', { current });
			} else if (target.config.restart === 'always') {
				// Otherwise only start the container if wants to be restarted
				return generateStep('start', { target });
			}
```
to avoid having to check `current.config.running` twice

Reply:
great optimization @Page-. I might remove the restart condition check entirely as some new questions have come to mind. If I leave it I'll definitely add this change.